File: Java/src/HeapSort.java
import java.io.*;
import java.util.*;

class HeapSort {

    public static void main(String[] args) throws Exception {
        long start = System.currentTimeMillis();
        //read file
        File file = new File("D:/KULIAH SMT 3/DAA/JURNAL/daa/src/nanda_wordlist.txt");
        Scanner sc = new Scanner(file);
        ArrayList<String> list = new ArrayList<>();
        while (sc.hasNextLine()) {
            list.add(sc.nextLine());
        }
        sc.close();
        //sort
        String[] arr = list.toArray(new String[0]);
        heapSort(arr);
        //write file
        FileWriter writer = new FileWriter("Heap_Java.txt");
        for (String str : arr) {
            writer.write(str + System.lineSeparator());
        }
        writer.close();
        long time = System.currentTimeMillis() - start;
        System.out.println("lama waktu (TimeMillis): " + time);
    }

    public static void heapSort(String[] arr) {
        int n = arr.length;
        for (int i = n / 2 - 1; i >= 0; i--) {
            heapify(arr, n, i);
        }
        for (int i = n - 1; i >= 0; i--) {
            String temp = arr[0];
            arr[0] = arr[i];
            arr[i] = temp;
            heapify(arr, i, 0);
        }
    }

    public static void heapify(String[] arr, int n, int i) {
        int largest = i;
        int l = 2 * i + 1;
        int r = 2 * i + 2;
        if (l < n && arr[l].compareTo(arr[largest]) > 0) {
            largest = l;
        }
        if (r < n && arr[r].compareTo(arr[largest]) > 0) {
            largest = r;
        }
        if (largest != i) {
            String swap = arr[i];
            arr[i] = arr[largest];
            arr[largest] = swap;
            heapify(arr, n, largest);
        }
    }
}

File: Python/HeapSort.py
import time
awal = time.time()

def HeapSort(val):
    for start in range((len(val)-2)//2, -1, -1):
        siftdown(val, start, len(val)-1)

    for end in range(len(val)-1, 0, -1):
        val[end], val[0] = val[0], val[end]
        siftdown(val, 0, end - 1)
    return val

def siftdown(val, start, end):
    root = start
    while True:
        child = root * 2 + 1
        if child > end: break
        if child + 1 <= end and val[child] < val[child + 1]:
            child += 1
        if val[root] < val[child]:
            val[root], val[child] = val[child], val[root]
            root = child
        else:
            break
        
words = []
infile = open("nanda_wordlist.txt","r")
for line in infile:
    temp = line.split()
    for i in temp:
      words.append(i)
infile.close()

HeapSort(words)
outfile = open("HeapSort_PY.txt", "w")
for i in words:
    outfile.writelines(i)
    outfile.writelines(" ")
outfile.close()

print("Running Time : %s detik" % (time.time() - awal))
